_source/_posts/2023-07-27-spring-graphql-react.adoc

```diff
@@ -0,0 +1,1857 @@
+
+[source, shell]
+----
+npm run dev
+----
+
+Navigate to http://localhost:3000 and you should see the companies list:
+
+image::{% asset_path blog/spring-graphql-react/react-datagrid.png %}[alt=Home page companies datagrid,width=900,align=center]
+
+== Add security with Auth0
+
+For securing both the server and client, the Auth0 platform provides the best customer experience, and with a few simple configuration steps, you can add authentication to your applications. Sign up at https://auth0.com/signup[Auth0] and install the https://github.com/auth0/auth0-cli[Auth0 CLI] that will help you create the tenant and the client applications.
+
+=== Add resource server security to the GraphQL API server
+
+In the command line login to Auth0 with the CLI:
+
+[source, shell]
+----
+auth0 login
+----
+
+The command output will display a device confirmation code and open a browser session to activate the device.
+
+**NOTE**: My browser was not displaying anything, so I had to manually activate the device by opening the URL `https://auth0.auth0.com/activate?user_code={deviceCode}`.
+
+On successful login, you will see the tenant, which you will use as the issuer later:
+
+----
+✪ Welcome to the Auth0 CLI 🎊
+
+If you don't have an account, please create one here: https://auth0.com/signup.
+
+Your device confirmation code is: KGFL-LNVB
+
+ ▸    Press Enter to open the browser to log in or ^C to quit...
+
+Waiting for the login to complete in the browser... ⣻Opening in existing browser session.
+Waiting for the login to complete in the browser... done
+
+ ▸    Successfully logged in.
+ ▸    Tenant: dev-avup2laz.us.auth0.com
+----
+
+The next step is to create a client app, which you can do in one command:
+
+[source, shell]
+----
+auth0 apps create \
+  --name "GraphQL Server" \
+  --description "Spring Boot GraphQL Resource Server" \
+  --type regular \
+  --callbacks http://localhost:8080/login/oauth2/code/okta \
+  --logout-urls http://localhost:8080 \
+  --reveal-secrets
+----
+
+Once the app is created, you will see the OIDC app's configuration:
+
+----
+=== dev-avup2laz.us.auth0.com application created
+
+  CLIENT ID            ***
+  NAME                 GraphQL Server
+  DESCRIPTION          Spring Boot GraphQL Resource Server
+  TYPE                 Regular Web Application
+  CLIENT SECRET        ***
+  CALLBACKS            http://localhost:8080/login/oauth2/code/okta
+  ALLOWED LOGOUT URLS  http://localhost:8080
+  ALLOWED ORIGINS
+  ALLOWED WEB ORIGINS
+  TOKEN ENDPOINT AUTH
+  GRANTS               implicit, authorization_code, refresh_token, client_credentials
+
+ ▸    Quickstarts: https://auth0.com/docs/quickstart/webapp
+ ▸    Hint: Emulate this app's login flow by running `auth0 test login ***`
+ ▸    Hint: Consider running `auth0 quickstarts download ***`
+----
+
+Add the `okta-spring-boot-starter` dependency to the `build.gradle` file:
+----
+dependencies {
+    ...
+  implementation 'com.okta.spring:okta-spring-boot-starter:3.0.4'
```

Comment:
Please indent ... and implementation so it matches the build.gradle, which I believe is 4 spaces. At the very least, these lines should left-align.

Reply:
Fixed indentation